chall09/edramire.c

```diff
@@ -0,0 +1,157 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   edramire.c                                         :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: edramire <edramire@student.42.fr>          +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2020/05/09 15:19:48 by edramire          #+#    #+#             */
+/*   Updated: 2020/05/09 16:59:3 by edramire         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include <string.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <ctype.h>
+#include <limits.h>
+#include <stdio.h>
+
+#define MAX_SECONDS 10000000000000000000ULL
+#define DAYS_SECONDS 86400
+#define HOURS_SECONDS 3600
+#define YEAR_SECONDS 31536000
+
+
+typedef struct	s_time
+{
+	unsigned long long int	years;
+	unsigned int			days;
+	unsigned int			hours;
+	unsigned int			minutes;
+	unsigned int			seconds;
+	unsigned int			flags;
+}				t_time;
+
+static char		ft_stc_validate_duration(char *seconds)
+{
+	
+	if (seconds == NULL || strlen(seconds) == 0)
+		return (1);
+	while (*seconds != '\0')
+	{
+		if (isdigit(*seconds) > 0)
+			break;
+		else if (*seconds++ != ' ')
+			return (1);
+	}
+	while (*seconds != '\0')
+	{
+		if (*seconds == ' ')
+			break;
+		else if (isdigit(*seconds++) == 0)
+			return (1);
+	}
+	while (*seconds != '\0')
+	{
+		if (*seconds++ != ' ')
+			return (1);
+	}
+	return (0);
+}
+
+static unsigned long long int	ft_stc_convert_number(char *seconds)
+{
+	unsigned long long int	number;
+	int						len;
+
+	number = 0;
+	len = 0;
+	while (*seconds != '\0' && isdigit(*seconds) == 0)
+		++seconds;
+	while (*seconds != '\0' && isdigit(*seconds) > 0 && len < 20 )
+	{
+		number = number * 10 + (*seconds++ - '0');
+		len += len == 0 && number == 0 ? 0 : 1;
+	}
+	if (len == 20 && (number != MAX_SECONDS || isdigit(*seconds) > 0))
+		return (ULLONG_MAX);
+	return (number);
+}
+
+static t_time	ft_stc_set_flags(t_time time)
+{
+	if (time.years > 0)
+		time.flags |= 0x10;
+	if (time.days > 0)
+		time.flags |= 0x8;
+	if (time.hours > 0)
+		time.flags |= 0x4;
+	if (time.minutes > 0)
+		time.flags |= 0x2;
+	if (time.seconds > 0)
+		time.flags |= 0x1;
+	return (time);
+}
+static void		ft_stc_format_time(char *str, char const *txt, int prev,
+					int next, unsigned long long int time)
+{
+	char	*aux;
+
+	if (time == 0)
+		return ;
+	if (prev > 0)
+	{
+		aux = strdup(str);
+		if (aux == NULL)
+			return ;
```

Comment:
There should be a way to notify the caller that the malloc failed.  
Returning a code, or setting `str` to null (you would have to change it to a higher level pointer)